src/cpu/dyn_cache.h

```diff
@@ -566,28 +567,22 @@ CacheBlock::~CacheBlock() {
 }
 
 void CacheBlock::DeleteWriteMask() {
-	delete[] cache.wmapmask;
-	cache.wmapmask = {};
-
-	cache.masklen = 0;
+	if (cache.wmapmask) {
+		delete cache.wmapmask;
+		cache.wmapmask = nullptr;
+	}
 }
 
-void CacheBlock::GrowWriteMask(const uint16_t new_mask_len) {
+void CacheBlock::GrowWriteMask(const uint16_t new_mask_len)
+{
 	// This function is only called to increase the mask
-	assert(new_mask_len > cache.masklen);
-
-	// Allocate the new mask
-	auto new_mask = new uint8_t[new_mask_len];
-	memset(new_mask, 0, new_mask_len);
-
-	// Copy the current into the new
-	std::copy(cache.wmapmask, cache.wmapmask + cache.masklen, new_mask);
-
-	// Update the current
-	delete[] cache.wmapmask;
-	cache.wmapmask = new_mask;
-
-	cache.masklen = new_mask_len;
+	if (cache.wmapmask) {
+		assert(new_mask_len > cache.wmapmask->size());
+		cache.wmapmask->resize(new_mask_len);
```

Comment:
Do we know what are the expected / maximum sizes of `wmapmask`? We could put a `wmapmask->reserve(…)` call somewhere to ensure the C++ standard library won’t move the vector content in memory (maybe it doesn’t already - implementations are smart enough to not always return the extra memory if vector is being shrink - but I would prefer to be sure).

Reply:
Yup; it's capped at 65k pages - the cache of dynamically-generated pages caps out around 256 MB.

> If not, definitely don't use a vector, just store the length manually and use a a 8-byte long uint8_t buffer.

The good news is the `std::vector` is only has three members: the pointer, capacity, and current size (it avoids re-allocating when when `resize(n)`ing < `capacity`).